### PR TITLE
REGR: SpooledTemporaryFile support in read_csv

### LINKED ISSUE #43439: BUG: SpooledTemporaryFile no longer working with read_csv
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import os
import pandas
import tempfile
import traceback


with tempfile.SpooledTemporaryFile() as f:
    f.write(b'abcd')
    f.seek(0)

    try:
        result = pandas.read_csv(f)
        print(result)

    except Exception as e:
        print(e)
        print(traceback.format_exc())
```

### Issue Description

This worked in pandas 1.1.5 but is no longer working in version >=1.2.0.

### Expected Behavior

According to the `read_csv` description I would expect this to work.

filepath_or_bufferstr, path object or file-like object

> Any valid string path is acceptable. The string could be a URL. Valid URL schemes include http, ftp, s3, gs, and file. For file URLs, a host is expected. A local file could be: file://localhost/path/to/table.csv.
> If you want to pass in a path object, pandas accepts any os.PathLike.
> By file-like object, we refer to objects with a read() method, such as a file handle (e.g. via builtin open function) or StringIO.

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 5f648bf1706dd75a9ca0d29f26eadfbb595fe52b
python           : 3.7.9.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.5.0
Version          : Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : None
LOCALE           : en_US.UTF-8

pandas           : 1.3.2
numpy            : 1.21.2
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 20.1.1
setuptools       : 47.1.0
IPython          : 7.27.0
</details>

### PR DESCRIPTION
**Pull request by twoertwein**
- [ ] closes #43439
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Probably should throw a warning, if we cannot wrap the handle with `io.TextIOWrapper`.

To avoid the try-except, we could check whether the handle has the `readable` attribute, but there might be other attribute errors that might occur inside `TextIOWrapper`.

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -723,20 +723,25 @@ def get_handle(
...
+            handle = TextIOWrapper(
+                # error: Argument 1 to "TextIOWrapper" has incompatible type
+                # "Union[IO[bytes], IO[Any], RawIOBase, BufferedIOBase, TextIOBase, mmap]";
+                # expected "IO[bytes]"
+                handle,  # type: ignore[arg-type]
+                encoding=ioargs.encoding,
+                errors=errors,
+                newline="",
+            )
+            handles.append(handle)
```
Comment (jbrockmendel) [inline pandas/io/common.py:737]: can this be moved outside the try/except?

**Comment by simonjayhawkins** [comment]
Thanks @twoertwein. Is this targeting 1.3.x?

**Comment by twoertwein** [comment]
> Thanks @twoertwein. Is this targeting 1.3.x?

It should be, I'll probably have time this evening to finish it up.

I'm not yet sure whether try-except is a good solution here. Another option (maybe more for 1.4), would be to add a new keyword to read_csv to set the mode (to_csv has that currently). Then a user could overwrite the binary/text auto-detection by specifying `mode="rt"` or `mode="rb"`.

**Comment by twoertwein** [comment]
Instead of using a try-except block, I added `tempfile.SpooledTemporaryFile` to the list of classes that should always be treated as text handles.

I'm not sure whether #43439 is a regression: yes, it worked previously, but it only worked because of inconsistencies between the c-engine and python-engine. This PR allows this inconsistency to continue (I'm not a fan of this). It is great that the c-engine is not only faster than the python-engine, but also doesn't seem to need the`io.TextIOWrapper` (skipping the wrapper for the c-engine might actually give a slight performance improvement).

I would be slightly in favor of not fixing #43439 for 1.3.3 and instead adding a new keyword argument ("mode", same as in `to_csv`) in 1.4 for `read_csv` which allows the user to over-write the text/binary auto-detection. (And a PR for 1.4 to make the c-engine "dumber": if it is told to get strings, but it gets bytes, it should fail.).

If it is preferred to fix #43439 for 1.3.3, I'm happy to make this happen (I think this PR should be ready).

**Comment by jreback** [comment]
oh new commit way nicer

**Comment by jreback** [comment]
very nice @twoertwein

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**